src/rgw/rgw_rest_s3.h

```diff
@@ -645,7 +645,48 @@ static inline int valid_s3_bucket_name(const string& name, bool relaxed=false)
   return 0;
 }
 
-class RGWLDAPAuthEngine: RGWTokenBasedAuthEngine
+class RGWS3V2AuthEngine : public RGWAuthEngine {
+protected:
+  std::string access_key_id;
+  std::string signature;
+  std::string expires;
+  bool qsr;
+
+public:
+  class Extractor {
+  public:
+    virtual ~Extractor() {};
+    virtual void get_auth_keys(std::string& access_key_id,
+                            std::string& signature,
+                            std::string& expires,
+                            bool& qsr) const = 0;
+  };
+
+  RGWS3V2AuthEngine(CephContext* const cct, const Extractor& extr)
+    : RGWAuthEngine(cct) {
+    extr.get_auth_keys(access_key_id, signature, expires, qsr);
+  }
+
+  bool is_applicable() const noexcept override {
+    return ! (access_key_id.empty() && signature.empty());
+  }
+};
+
+class RGWS3V2Extractor : public RGWS3V2AuthEngine::Extractor {
+protected:
+  const req_state* const s;
+
+public:
+  RGWS3V2Extractor(const req_state * const s)
+    : s(s) {}
+
+  void get_auth_keys(std::string& access_key_id,
+                  std::string& signature,
+                  std::string& expires,
+                  bool& qsr) const;
```

Comment:
`override` would be very helpful here. Also indentation of parameters might be improved but this is less than minor.